database/src/service/write.rs

```diff
@@ -226,5 +229,26 @@ impl DatabaseWriter {
 /// [`WriteRequest::WriteBlock`].
 #[inline]
 fn write_block(env: &ConcreteEnv, block: &VerifiedBlockInformation) -> ResponseResult {
-    todo!()
+    let env_inner = env.env_inner();
+    let tx_rw = env_inner.tx_rw()?;
+
+    let result = {
+        let mut tables_mut = env_inner.open_tables_mut(&tx_rw)?;
+        crate::ops::block::add_block(block, &mut tables_mut)
+    };
+
+    match result {
+        Ok(()) => {
+            tx_rw.commit()?;
+            Ok(Response::WriteBlockOk)
+        }
+        Err(e) => {
+            // INVARIANT: ensure database atomicity by aborting
+            // the transaction on `add_block()` failures.
+            tx_rw
+                .abort()
+                .expect("could not maintain database atomicity by aborting write transaction");
+            Err(e)
+        }
+    }
 }
```

Comment:
If adding a block fails we should panic if we cannot maintain atomicity, correct?

Otherwise the database may proceed with half-inserted data.

Reply:
yes, however does failing to abort cause data to half write?

Reply:
- `heed`'s abort is infallible and panics if the tx was already commited (API doesn't let us do this so it's unreachable)
- `redb`'s abort is fallible: https://docs.rs/redb/2.0.0/src/redb/transactions.rs.html#939

The above eventually calls https://docs.rs/redb/2.0.0/src/redb/tree_store/page_store/page_manager.rs.html#600-642 which always returns `Ok(())` or panics. So, regardless we will panic if data cannot be rolled back. I think we should keep panics here too as to rely on the signature and not the impl.